src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -218,7 +219,27 @@ internal async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool
                             targetsList = await GetTargetsListAsync(assetsFilePath, token);
                         }
 
-                        transitiveOrigins = calculatedTransitivePackages.Any() ? ComputeTransitivePackageOrigins(calculatedInstalledPackages, targetsList, token) : new Dictionary<string, TransitiveEntry>();
+                        List<LockFileTargetLibrary> projectReferences = targetsList.SelectMany(t => t.Libraries).Where(l => l.Type == LibraryType.Project).ToList();
+                        List<PackageReference> installedProjectReferences = new List<PackageReference>(projectReferences.Count());
+
+                        // Since ProjectReferences are not in the dependencies section of the package spec (installed packages), we need to manually add them
+                        // to the installed packages list to compute transitive origins
+                        foreach (var projectReference in projectReferences)
+                        {
+                            var packageReference = new PackageReference(
+                                identity: new PackageIdentity(projectReference.Name, projectReference.Version),
+                                targetFramework: new NuGetFramework(projectReference.Framework),
+                                userInstalled: false,
+                                developmentDependency: false,
+                                requireReinstallation: false,
+                                allowedVersions: VersionRange.Parse(projectReference.Version.ToString()));
+                            installedProjectReferences.Add(packageReference);
+                        }
+
+                        installedProjectReferences.AddRange(calculatedInstalledPackages);
```

Comment:
It seems strange to add calculatedInstalledPackages to a list called installedProjectReferences since these calculatedInstalledPackages are packages, not projects. We should rename the variable but this also indicates to me that we're mixing the concept of projects and packages here and treating projects as packages which may limit future changes.

Reply:
So `calculatedInstalledPackages` is calculated by looking at the the dgspec file, (calculated from the .csproj) which has different nodes for projects and packages, and then we compare this file against the assets file (which has the same nodes for projects and packages). 

That's also why I had to remove type validation in https://github.com/NuGet/NuGet.Client/pull/5926/files#diff-cf69ebf676e61580349108d38e5dab1bbff91468194c6a8d342149d264139b15L326. Removing just that line makes the PM UI start to render `ProjectReferences` as transitive packages and that is also why added a check for TransitiveOrigins cache in the UI to avoid rendering them https://github.com/NuGet/NuGet.Client/blob/38ab39a6691e6e5d54e29055d9e258b8b5d57220/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs#L1038

Reply:
It just occurred to me that this part of PM UI can be simplified.  One part of the assets file contains a flat list of packages (and projects), and another part of the assets file contains a list of direct referenced by the project. There's no need for graph walking.

Something like:
```
foreach (var package in assetsFile.Targets.Where(node => node.Type == "package"))
{
   bool isTransitive = !assetsFile.Package.Dependencies.Contains(package.Id);
}
```

Graph walking is only needed when calculating the transitive root, or when we eventually show a tree representing the graph.

Reply:
Looks like we were limiting the assets file reading to the Targets section, `Libraries` is the one you are mentioning. There is even more information about Project References in `ProjectFileDependencyGroup`. 

Even if we are using Targets section, I believe we shouldn't be calculating the whole graph

Reply:
@zivkan With that suggestion I was able to remove several graph iterations and removed nodes that weren't packages.